File: backend/src/main/java/com/atomix/cafeteria/service/VendorService.java
package com.atomix.cafeteria.service;

import com.atomix.cafeteria.dto.*;
import com.atomix.cafeteria.entity.Vendor;
import com.atomix.cafeteria.entity.VendorStatus;
import com.atomix.cafeteria.entity.VendorType;
import com.atomix.cafeteria.repository.VendorRepository;
import org.modelmapper.ModelMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Transactional
public class VendorService {
    
    private static final Logger logger = LoggerFactory.getLogger(VendorService.class);
    
    private final VendorRepository vendorRepository;
    private final ModelMapper modelMapper;
    
    @Autowired
    public VendorService(VendorRepository vendorRepository, ModelMapper modelMapper) {
        this.vendorRepository = vendorRepository;
        this.modelMapper = modelMapper;
    }
    
    /**
     * Create a new vendor
     */
    public VendorResponse createVendor(VendorRequest request) {
        logger.info("Creating new vendor: {}", request.getName());
        logger.debug("Vendor request details: name={}, email={}, type={}", 
                    request.getName(), request.getContactEmail(), request.getVendorType());
        
        try {
            // Validate request
            validateVendorRequest(request);
            logger.debug("Vendor request validation passed");
            
            // Check for duplicate name or email
            if (vendorRepository.findByName(request.getName()).isPresent()) {
                logger.warn("Vendor creation failed: Name '{}' already exists", request.getName());
                throw new RuntimeException("Vendor with name '" + request.getName() + "' already exists");
            }
            
            if (vendorRepository.findByContactEmail(request.getContactEmail()).isPresent()) {
                logger.warn("Vendor creation failed: Email '{}' already exists", request.getContactEmail());
                throw new RuntimeException("Vendor with email '" + request.getContactEmail() + "' already exists");
            }
            
            logger.debug("Duplicate check passed, proceeding with entity creation");
            
            // Create vendor entity
            Vendor vendor = modelMapper.map(request, Vendor.class);
            vendor.setStatus(VendorStatus.PENDING);
            vendor.setIsActive(true);
            vendor.setAverageRating(0.0);
            vendor.setTotalReviews(0);
            
            logger.debug("Vendor entity created, saving to database");
            
            // Save vendor
            vendor = vendorRepository.save(vendor);
            
            logger.info("Successfully created vendor with ID: {}", vendor.getId());
            return mapToVendorResponse(vendor);
            
        } catch (RuntimeException e) {
            logger.error("Runtime error creating vendor: {}", e.getMessage());
            throw e;
        } catch (Exception e) {
            logger.error("Unexpected error creating vendor: ", e);
            throw new RuntimeException("Failed to create vendor: " + e.getMessage(), e);
        }
    }
    
    /**
     * Get vendor by ID
     */
    @Transactional(readOnly = true)
    public VendorResponse getVendorById(Long id) {
        Vendor vendor = vendorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + id));
        
        return mapToVendorResponse(vendor);
    }
    
    /**
     * Get all vendors with pagination
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getAllVendors(int page, int size, String sortBy, String sortDir) {
        Sort sort = Sort.by(Sort.Direction.fromString(sortDir), sortBy);
        Pageable pageable = PageRequest.of(page, size, sort);
        
        Page<Vendor> vendors = vendorRepository.findAll(pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get vendors by status
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getVendorsByStatus(VendorStatus status, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<Vendor> vendors = vendorRepository.findByStatus(status, pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get currently active vendors
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getCurrentlyActiveVendors(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Vendor> vendors = vendorRepository.findCurrentlyActiveVendors(LocalDateTime.now(), pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get vendors by type
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getVendorsByType(VendorType type, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("name"));
        Page<Vendor> vendors = vendorRepository.findByVendorType(type, pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get vendors by floor
     */
    @Transactional(readOnly = true)
    public List<VendorResponse> getVendorsByFloor(String floorId, boolean activeOnly) {
        List<Vendor> vendors;
        if (activeOnly) {
            vendors = vendorRepository.findActiveVendorsByFloorId(floorId);
        } else {
            vendors = vendorRepository.findByFloorId(floorId);
        }
        
        return vendors.stream()
            .map(this::mapToVendorResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Search vendors
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> searchVendors(String searchTerm, boolean activeOnly, int page, int size) {
        if (activeOnly) {
            Pageable pageable = PageRequest.of(page, size, Sort.by("averageRating").descending());
            Page<Vendor> vendors = vendorRepository.searchActiveVendors(searchTerm, pageable);
            return vendors.map(this::mapToVendorResponse);
        } else {
            List<Vendor> vendors = vendorRepository.searchVendors(searchTerm);
            // Convert to Page manually for consistency
            int start = (int) PageRequest.of(page, size).getOffset();
            int end = Math.min((start + size), vendors.size());
            List<Vendor> subList = vendors.subList(start, end);
            
            return new org.springframework.data.domain.PageImpl<>(
                subList.stream().map(this::mapToVendorResponse).collect(Collectors.toList()),
                PageRequest.of(page, size),
                vendors.size()
            );
        }
    }
    
    /**
     * Update vendor
     */
    public VendorResponse updateVendor(Long id, VendorRequest request) {
        logger.info("Updating vendor with ID: {}", id);
        
        Vendor vendor = vendorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + id));
        
        // Validate request
        validateVendorRequest(request);
        
        // Check for duplicate name or email (excluding current vendor)
        Optional<Vendor> existingNameVendor = vendorRepository.findByName(request.getName());
        if (existingNameVendor.isPresent() && !existingNameVendor.get().getId().equals(id)) {
            throw new RuntimeException("Vendor with name '" + request.getName() + "' already exists");
        }
        
        Optional<Vendor> existingEmailVendor = vendorRepository.findByContactEmail(request.getContactEmail());
        if (existingEmailVendor.isPresent() && !existingEmailVendor.get().getId().equals(id)) {
            throw new RuntimeException("Vendor with email '" + request.getContactEmail() + "' already exists");
        }
        
        // Update vendor fields
        vendor.setName(request.getName());
        vendor.setDescription(request.getDescription());
        vendor.setContactEmail(request.getContactEmail());
        vendor.setContactPhone(request.getContactPhone());
        vendor.setContactPerson(request.getContactPerson());
        vendor.setBusinessLicense(request.getBusinessLicense());
        vendor.setLogoUrl(request.getLogoUrl());
        vendor.setOperatingHours(request.getOperatingHours());
        vendor.setLocationDescription(request.getLocationDescription());
        vendor.setFloorIds(request.getFloorIds());
        vendor.setVendorType(request.getVendorType());
        vendor.setTemporaryStartDate(request.getTemporaryStartDate());
        vendor.setTemporaryEndDate(request.getTemporaryEndDate());
        
        vendor = vendorRepository.save(vendor);
        
        logger.info("Updated vendor with ID: {}", id);
        return mapToVendorResponse(vendor);
    }
    
    /**
     * Update vendor status
     */
    public VendorResponse updateVendorStatus(Long id, VendorStatusUpdateRequest request) {
        logger.info("Updating vendor status for ID: {} to {}", id, request.getStatus());
        
        Vendor vendor = vendorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + id));
        
        VendorStatus oldStatus = vendor.getStatus();
        vendor.setStatus(request.getStatus());
        
        vendor = vendorRepository.save(vendor);
        
        logger.info("Updated vendor status from {} to {} for vendor ID: {}", 
                   oldStatus, request.getStatus(), id);
        
        return mapToVendorResponse(vendor);
    }
    
    /**
     * Toggle vendor active status
     */
    public VendorResponse toggleVendorActiveStatus(Long id) {
        logger.info("Toggling active status for vendor ID: {}", id);
        
        Vendor vendor = vendorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + id));
        
        vendor.setIsActive(!vendor.getIsActive());
        vendor = vendorRepository.save(vendor);
        
        logger.info("Toggled vendor active status to {} for vendor ID: {}", vendor.getIsActive(), id);
        return mapToVendorResponse(vendor);
    }
    
    /**
     * Delete vendor
     */
    public void deleteVendor(Long id) {
        logger.info("Deleting vendor with ID: {}", id);
        
        Vendor vendor = vendorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + id));
        
        // Check if vendor has associated menu items
        if (!vendor.getMenuItems().isEmpty()) {
            throw new RuntimeException("Cannot delete vendor with existing menu items. Please remove all menu items first.");
        }
        
        vendorRepository.delete(vendor);
        logger.info("Deleted vendor with ID: {}", id);
    }
    
    /**
     * Get pending approval vendors
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getPendingApprovalVendors(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Vendor> vendors = vendorRepository.findPendingApprovalVendors(pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get high-rated vendors
     */
    @Transactional(readOnly = true)
    public Page<VendorResponse> getHighRatedVendors(Double minRating, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Vendor> vendors = vendorRepository.findHighRatedActiveVendors(minRating, pageable);
        return vendors.map(this::mapToVendorResponse);
    }
    
    /**
     * Get vendor statistics
     */
    @Transactional(readOnly = true)
    public Map<String, Object> getVendorStatistics() {
        Map<String, Object> stats = new java.util.HashMap<>();
        
        stats.put("totalVendors", vendorRepository.count());
        stats.put("activeVendors", vendorRepository.countActiveVendors());
        stats.put("pendingApproval", vendorRepository.countByStatus(VendorStatus.PENDING));
        stats.put("approvedVendors", vendorRepository.countByStatus(VendorStatus.APPROVED));
        stats.put("rejectedVendors", vendorRepository.countByStatus(VendorStatus.REJECTED));
        stats.put("suspendedVendors", vendorRepository.countByStatus(VendorStatus.SUSPENDED));
        
        // Vendor type statistics
        List<Object[]> typeStats = vendorRepository.getVendorTypeStatistics();
        Map<String, Long> vendorTypeStats = typeStats.stream()
            .collect(Collectors.toMap(
                arr -> ((VendorType) arr[0]).getDisplayName(),
                arr -> (Long) arr[1]
            ));
        stats.put("vendorTypeStatistics", vendorTypeStats);
        
        return stats;
    }
    
    /**
     * Handle expired temporary vendors
     */
    @Transactional
    public void handleExpiredTemporaryVendors() {
        List<Vendor> expiredVendors = vendorRepository.findExpiredTemporaryVendors(LocalDateTime.now());
        
        for (Vendor vendor : expiredVendors) {
            vendor.setIsActive(false);
            vendor.setStatus(VendorStatus.INACTIVE);
            vendorRepository.save(vendor);
            logger.info("Marked expired temporary vendor as inactive: {}", vendor.getName());
        }
        
        if (!expiredVendors.isEmpty()) {
            logger.info("Processed {} expired temporary vendors", expiredVendors.size());
        }
    }
    
    /**
     * Update vendor rating
     */
    public VendorResponse updateVendorRating(Long vendorId, Double rating) {
        logger.info("Updating rating for vendor ID: {} with rating: {}", vendorId, rating);
        
        if (rating < 1.0 || rating > 5.0) {
            throw new RuntimeException("Rating must be between 1.0 and 5.0");
        }
        
        Vendor vendor = vendorRepository.findById(vendorId)
            .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + vendorId));
        
        vendor.updateRating(rating);
        vendor = vendorRepository.save(vendor);
        
        logger.info("Updated vendor rating. New average: {}, Total reviews: {}", 
                   vendor.getAverageRating(), vendor.getTotalReviews());
        
        return mapToVendorResponse(vendor);
    }
    
    // Private helper methods
    
    private void validateVendorRequest(VendorRequest request) {
        if (request.isTemporaryVendor() && !request.hasValidDateRange()) {
            throw new RuntimeException("Temporary vendors must have valid start and end dates");
        }
        
        if (request.isTemporaryVendor() && 
            request.getTemporaryStartDate() != null && 
            request.getTemporaryStartDate().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Temporary vendor start date cannot be in the past");
        }
    }
    
    private VendorResponse mapToVendorResponse(Vendor vendor) {
        VendorResponse response = modelMapper.map(vendor, VendorResponse.class);
        
        // Set computed fields
        response.setStatusDisplayName(vendor.getStatus().getDisplayName());
        response.setVendorTypeDisplayName(vendor.getVendorType().getDisplayName());
        response.setIsCurrentlyActive(vendor.isCurrentlyActive());
        response.setIsTemporary(vendor.isTemporary());
        response.setTotalMenuItems(vendor.getMenuItems().size());
        
        return response;
    }
} 

File: frontend/src/services/vendorService.ts
import axios from 'axios';

const API_BASE_URL = process.env.REACT_APP_API_BASE_URL || 'http://localhost:8083/api/v1';

// Types
export interface VendorType {
  PERMANENT: 'PERMANENT';
  TEMPORARY: 'TEMPORARY';
  SEASONAL: 'SEASONAL';
  EVENT_BASED: 'EVENT_BASED';
}

export interface VendorStatus {
  PENDING: 'PENDING';
  APPROVED: 'APPROVED';
  REJECTED: 'REJECTED';
  SUSPENDED: 'SUSPENDED';
  INACTIVE: 'INACTIVE';
}

export interface VendorRequest {
  name: string;
  description?: string;
  contactEmail: string;
  contactPhone?: string;
  contactPerson?: string;
  businessLicense?: string;
  logoUrl?: string;
  operatingHours?: string;
  locationDescription?: string;
  floorIds?: string[];
  vendorType: string;
  temporaryStartDate?: string;
  temporaryEndDate?: string;
}

export interface VendorResponse {
  id: number;
  name: string;
  description?: string;
  contactEmail: string;
  contactPhone?: string;
  contactPerson?: string;
  businessLicense?: string;
  logoUrl?: string;
  status: string;
  statusDisplayName: string;
  isActive: boolean;
  operatingHours?: string;
  locationDescription?: string;
  floorIds: string[];
  vendorType: string;
  vendorTypeDisplayName: string;
  temporaryStartDate?: string;
  temporaryEndDate?: string;
  averageRating: number;
  totalReviews: number;
  totalMenuItems: number;
  isCurrentlyActive: boolean;
  isTemporary: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface VendorStatusUpdateRequest {
  status: string;
  reason?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

export interface PaginatedResponse<T> {
  content: T[];
  pageable: {
    sort: {
      empty: boolean;
      sorted: boolean;
      unsorted: boolean;
    };
    offset: number;
    pageSize: number;
    pageNumber: number;
    unpaged: boolean;
    paged: boolean;
  };
  last: boolean;
  totalPages: number;
  totalElements: number;
  size: number;
  number: number;
  sort: {
    empty: boolean;
    sorted: boolean;
    unsorted: boolean;
  };
  first: boolean;
  numberOfElements: number;
  empty: boolean;
}

class VendorService {
  private readonly baseURL: string;

  constructor() {
    this.baseURL = `${API_BASE_URL}/vendors`;
  }

  // Get auth headers
  private getAuthHeaders() {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    };
  }

  // Create a new vendor
  async createVendor(vendorRequest: VendorRequest): Promise<VendorResponse> {
    try {
      console.log('Sending vendor creation request:', vendorRequest);
      console.log('Headers:', this.getAuthHeaders());
      console.log('URL:', this.baseURL);
      
      const response = await axios.post<ApiResponse<VendorResponse>>(
        this.baseURL,
        vendorRequest,
        { headers: this.getAuthHeaders() }
      );
      
      console.log('Vendor creation response:', response.data);
      return response.data.data;
    } catch (error: any) {
      console.error('Vendor creation error:', error);
      console.error('Error response:', error.response);
      console.error('Error status:', error.response?.status);
      console.error('Error data:', error.response?.data);
      
      // Enhanced error message extraction
      let errorMessage = 'Failed to create vendor';
      
      if (error.response?.status === 401) {
        errorMessage = 'Unauthorized - Please log in again';
      } else if (error.response?.status === 403) {
        errorMessage = 'Access denied - You need ADMIN or MANAGER role to create vendors';
      } else if (error.response?.data?.message) {
        errorMessage = error.response.data.message;
      } else if (error.response?.data?.data) {
        errorMessage = error.response.data.data;
      } else if (error.message) {
        errorMessage = error.message;
      }
      
      throw new Error(errorMessage);
    }
  }

  // Get vendor by ID
  async getVendorById(id: number): Promise<VendorResponse> {
    try {
      const response = await axios.get<ApiResponse<VendorResponse>>(
        `${this.baseURL}/${id}`,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendor');
    }
  }

  // Get all vendors with pagination
  async getAllVendors(
    page: number = 0,
    size: number = 10,
    sortBy: string = 'name',
    sortDir: string = 'asc'
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        this.baseURL,
        {
          params: { page, size, sortBy, sortDir },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendors');
    }
  }

  // Get vendors by status
  async getVendorsByStatus(
    status: string,
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/status/${status}`,
        {
          params: { page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendors by status');
    }
  }

  // Get currently active vendors
  async getCurrentlyActiveVendors(
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/active`,
        {
          params: { page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch active vendors');
    }
  }

  // Get vendors by type
  async getVendorsByType(
    type: string,
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/type/${type}`,
        {
          params: { page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendors by type');
    }
  }

  // Get vendors by floor
  async getVendorsByFloor(floorId: string, activeOnly: boolean = true): Promise<VendorResponse[]> {
    try {
      const response = await axios.get<ApiResponse<VendorResponse[]>>(
        `${this.baseURL}/floor/${floorId}`,
        {
          params: { activeOnly },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendors by floor');
    }
  }

  // Search vendors
  async searchVendors(
    query: string,
    activeOnly: boolean = true,
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/search`,
        {
          params: { q: query, activeOnly, page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to search vendors');
    }
  }

  // Update vendor
  async updateVendor(id: number, vendorRequest: VendorRequest): Promise<VendorResponse> {
    try {
      const response = await axios.put<ApiResponse<VendorResponse>>(
        `${this.baseURL}/${id}`,
        vendorRequest,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to update vendor');
    }
  }

  // Update vendor status
  async updateVendorStatus(
    id: number,
    statusRequest: VendorStatusUpdateRequest
  ): Promise<VendorResponse> {
    try {
      const response = await axios.patch<ApiResponse<VendorResponse>>(
        `${this.baseURL}/${id}/status`,
        statusRequest,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to update vendor status');
    }
  }

  // Toggle vendor active status
  async toggleVendorActiveStatus(id: number): Promise<VendorResponse> {
    try {
      const response = await axios.patch<ApiResponse<VendorResponse>>(
        `${this.baseURL}/${id}/toggle-active`,
        {},
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to toggle vendor status');
    }
  }

  // Delete vendor
  async deleteVendor(id: number): Promise<void> {
    try {
      await axios.delete<ApiResponse<void>>(
        `${this.baseURL}/${id}`,
        { headers: this.getAuthHeaders() }
      );
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to delete vendor');
    }
  }

  // Get pending approval vendors
  async getPendingApprovalVendors(
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/pending-approval`,
        {
          params: { page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch pending vendors');
    }
  }

  // Get high-rated vendors
  async getHighRatedVendors(
    minRating: number = 4.0,
    page: number = 0,
    size: number = 10
  ): Promise<PaginatedResponse<VendorResponse>> {
    try {
      const response = await axios.get<ApiResponse<PaginatedResponse<VendorResponse>>>(
        `${this.baseURL}/high-rated`,
        {
          params: { minRating, page, size },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch high-rated vendors');
    }
  }

  // Get vendor statistics
  async getVendorStatistics(): Promise<any> {
    try {
      const response = await axios.get<ApiResponse<any>>(
        `${this.baseURL}/statistics`,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendor statistics');
    }
  }

  // Update vendor rating
  async updateVendorRating(id: number, rating: number): Promise<VendorResponse> {
    try {
      const response = await axios.patch<ApiResponse<VendorResponse>>(
        `${this.baseURL}/${id}/rating`,
        null,
        {
          params: { rating },
          headers: this.getAuthHeaders(),
        }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to update vendor rating');
    }
  }

  // Get vendor types
  async getVendorTypes(): Promise<string[]> {
    try {
      const response = await axios.get<ApiResponse<string[]>>(
        `${this.baseURL}/types`,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendor types');
    }
  }

  // Get vendor statuses
  async getVendorStatuses(): Promise<string[]> {
    try {
      const response = await axios.get<ApiResponse<string[]>>(
        `${this.baseURL}/statuses`,
        { headers: this.getAuthHeaders() }
      );
      return response.data.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch vendor statuses');
    }
  }
}

const vendorService = new VendorService();
export default vendorService; 